From: intrigeri <intrigeri@boum.org>
Date: Sun, 6 Jan 2019 11:07:08 +0100
Subject: If SOURCE_DATE_EPOCH is set,
 also clamp content timestamps with that value.

Based on a patch by Alexander Couzens <lynxis@fe...> posted on
https://sourceforge.net/p/squashfs/mailman/message/34673610/

Updated by Chris Lamb <lamby@debian.org> to use (time_t)0 as a
sentinel value over -1.
---
 squashfs-tools/mksquashfs.c | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/squashfs-tools/mksquashfs.c b/squashfs-tools/mksquashfs.c
index 7a6a14c..62888ec 100644
--- a/squashfs-tools/mksquashfs.c
+++ b/squashfs-tools/mksquashfs.c
@@ -83,6 +83,7 @@ int fd;
 struct squashfs_super_block sBlk;
 
 time_t forced_time = (time_t)0;
+time_t content_clamp_time = (time_t)0;
 
 /* filesystem flags for building */
 int comp_opts = FALSE;
@@ -2268,6 +2269,8 @@ restat:
 			pathname_reader(dir_ent), strerror(errno));
 		goto read_err;
 	}
+	if(content_clamp_time && buf2.st_mtime >= content_clamp_time)
+		buf2.st_mtime = content_clamp_time;
 
 	if(read_size != buf2.st_size) {
 		close(file);
@@ -3142,7 +3145,7 @@ void dir_scan(squashfs_inode *inode, char *pathname,
 		buf.st_mode = S_IRWXU | S_IRWXG | S_IRWXO | S_IFDIR;
 		buf.st_uid = getuid();
 		buf.st_gid = getgid();
-		buf.st_mtime = time(NULL);
+		buf.st_mtime = content_clamp_time ? content_clamp_time : time(NULL);
 		buf.st_dev = 0;
 		buf.st_ino = 0;
 		dir_ent->inode = lookup_inode2(&buf, PSEUDO_FILE_OTHER, 0);
@@ -3151,6 +3154,8 @@ void dir_scan(squashfs_inode *inode, char *pathname,
 			/* source directory has disappeared? */
 			BAD_ERROR("Cannot stat source directory %s because %s\n",
 				pathname, strerror(errno));
+		if(content_clamp_time && buf.st_mtime >= content_clamp_time)
+			buf.st_mtime = content_clamp_time;
 		dir_ent->inode = lookup_inode(&buf);
 	}
 
@@ -3373,6 +3378,8 @@ struct dir_info *dir_scan1(char *filename, char *subpath,
 			free_dir_entry(dir_ent);
 			continue;
 		}
+		if(content_clamp_time && buf.st_mtime >= content_clamp_time)
+			buf.st_mtime = content_clamp_time;
 
 		if((buf.st_mode & S_IFMT) != S_IFREG &&
 					(buf.st_mode & S_IFMT) != S_IFDIR &&
@@ -3529,7 +3536,7 @@ void dir_scan2(struct dir_info *dir, struct pseudo *pseudo)
 		buf.st_gid = pseudo_ent->dev->gid;
 		buf.st_rdev = makedev(pseudo_ent->dev->major,
 			pseudo_ent->dev->minor);
-		buf.st_mtime = time(NULL);
+		buf.st_mtime = content_clamp_time ? content_clamp_time : time(NULL);
 		buf.st_ino = pseudo_ino ++;
 
 		if(pseudo_ent->dev->type == 'd') {
@@ -5515,7 +5522,7 @@ printOptions:
 				"%lu but was found to be: %llu \n", ULONG_MAX, epoch);
 			EXIT_MKSQUASHFS();
 		}
-		forced_time = (time_t)epoch;
+		forced_time = content_clamp_time = (time_t)epoch;
 	}
 
 	/*
